DesignSafe-CI/portal
# [Job History] Implement expected functionality of "Reuse Inputs"

## Overview: ##
Implements reuse inputs functionality.
Any re-use has to use merge job data with defaults in app form.
There are two approaches of sharing job data with app form.
1. Url param: 
    * pass job uuid in parameter 
    * navigate to app form.
    * App form fetches job data.
    * With this the uuid is visible in the url.
2. Context:  
     * build a context and provider, 
     * wrap the workspace with the provider.
     * In jobs detail and modal, set the context with job data. And navigate to appform.
     * in appform, read job data from context.

I choose 1, because it is visible on what data the appform is working with - url shows the job uuid.


## PR Status: ##


## Related Jira tickets: ##

* [DES-2843](https://tacc-main.atlassian.net/browse/DES-2843)

## Summary of Changes: ##
* Implement merging of data
* add useJob which is disabled when uuid is not available.
* setup navigation from job to app form.

## Testing Steps: ##
1. Click 'Reuse Inputs' button from Job Listing. Ensure all the related inputs, parameters and configuration is populated.
4. Repeat the above test with 'Reuse Inputs' from Job Detail Modal.
5. Use a bad uuid in url: https://designsafe.dev/rw/workspace/FigureGen-Serial?appVersion=51.0.0&jobUUID=foobar and ensure the app is not broken.
6. Ensure existing app is not broken (no job uuid): https://designsafe.dev/rw/workspace/FigureGen-Serial?appVersion=51.0.0 
7. Ensure reuse button is shown for non-interactive jobs only
8. Ensure reuse button is not shown when job is in progress.

## UI Photos:
Reuse inputs :
<img width="427" alt="Screenshot 2024-07-15 at 10 36 59 AM" src="https://github.com/user-attachments/assets/bf5177ea-ace8-47a4-b5b3-01acf858057c">



<img width="1497" alt="Screenshot 2024-07-15 at 10 18 28 AM" src="https://github.com/user-attachments/assets/61cc645e-fdf7-48e2-a389-8d34d5293117">



Buttons not visible when job is in progress:
<img width="846" alt="Screenshot 2024-07-15 at 10 19 29 AM" src="https://github.com/user-attachments/assets/19221f0f-9704-4066-9dab-5c88a823922b">
<img width="334" alt="Screenshot 2024-07-15 at 10 19 25 AM" src="https://github.com/user-attachments/assets/ac6b5049-cf34-4e29-9957-294a70fcd8e2">





## Notes: ##
